File: sw/senetEthics.01/yaml/transformers/yamlPl.pl
%%% yaml to prolog transformer 
% by brygg ullmer (clemson) and copilot
% begun 2026-03-24

:- use_module(library(yaml)).

:- op(950, xfx, :=).

%%%%%%%% load_yaml(+File, -Data) %%%%%%%% 
load_yaml(File, Data) :-
    yaml_read(File, Data, [canonical(false)]).

%%%%%%%% extract_theme_abbrevs(+YAML, -Pairs) %%%%%%%% 
%% Produces pairs CategoryKey-[ThemeKeys...]

extract_theme_abbrevs(YAML, Pairs) :-
    Categories = YAML.categories,
    findall(Key-Themes,
        (   get_dict(Key, Categories, CatObj),
            get_dict(themes, CatObj, ThemeDict),
            dict_keys(ThemeDict, Themes)
        ),
        Pairs).

%%%%%%%% print_prolog_dsl(+Pairs) %%%%%%%% 
%% Emits lines like:
%%   themeAbbrev(civGovern) := [gov, civ, colAc, infr].

print_prolog_dsl([]).
print_prolog_dsl([Key-Themes | Rest]) :-
    format('themeAbbrev(~w) := ~w.~n', [Key, Themes]),
    print_prolog_dsl(Rest).

%%%%%%%% top-level convenience %%%%%%%% 
yaml_to_prolog_dsl(File) :-
    load_yaml(File, YAML),
    extract_theme_abbrevs(YAML, Pairs),
    print_prolog_dsl(Pairs).

%%% end %%%

File: sw/senetEthics.01/yaml/transformers/yamlPl.py
### yaml to prolog transformer 
# by brygg ullmer (clemson) and copilot
# begun 2026-03-24

import yaml

# Convert Python list to Prolog list syntax
def emit_theme_abbrev(key, themes):
  items = ", ".join(themes)
  return f"themeAbbrev({key}) := [{items}]."

def yaml_to_prolog_dsl(path):
  with open(path, "r") as f:
    data = yaml.safe_load(f)

  cats = data["categories"]
  lines = []

  for key, obj in cats.items():
    themes = list(obj["themes"].keys())
    lines.append(emit_theme_abbrev(key, themes))

  return "\n".join(lines)

if __name__ == "__main__":
  print(yaml_to_prolog_dsl("figures_themes03.yaml"))

### end ###
